File: src/main/java/org/move/ide/typing/MoveTextFieldTypedHandler.java
package org.move.ide.typing;

import com.intellij.codeInsight.AutoPopupController;
import com.intellij.codeInsight.completion.CompletionPhase;
import com.intellij.codeInsight.completion.impl.CompletionServiceImpl;
import com.intellij.codeInsight.editorActions.TypedHandlerDelegate;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiFile;
import org.jetbrains.annotations.NotNull;
import org.move.lang.core.psi.MvCodeFragment;

public class MoveTextFieldTypedHandler extends TypedHandlerDelegate {
    @NotNull
    @Override
    public Result checkAutoPopup(char charTyped, @NotNull Project project, @NotNull Editor editor, @NotNull PsiFile file) {
        if (!(file instanceof MvCodeFragment)) return Result.CONTINUE;

        var phase = CompletionServiceImpl.getCompletionPhase();
        if (charTyped == ':') {
            if (phase instanceof CompletionPhase.EmptyAutoPopup
                    && ((CompletionPhase.EmptyAutoPopup) phase).allowsSkippingNewAutoPopup(editor, charTyped)) {
                return Result.CONTINUE;
            }

            AutoPopupController.getInstance(project).scheduleAutoPopup(editor);
            return Result.STOP;
        }
        return Result.CONTINUE;
    }
}

File: src/main/kotlin/org/move/ide/typing/MoveTextFieldTypedHandler.kt
package org.move.ide.typing

import com.intellij.codeInsight.AutoPopupController
import com.intellij.codeInsight.completion.CompletionPhase.EmptyAutoPopup
import com.intellij.codeInsight.completion.impl.CompletionServiceImpl
import com.intellij.codeInsight.editorActions.TypedHandlerDelegate
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiFile
import org.move.lang.core.psi.MvCodeFragment

class MoveTextFieldTypedHandler : TypedHandlerDelegate() {
    override fun checkAutoPopup(charTyped: Char, project: Project, editor: Editor, file: PsiFile): Result {
        if (file !is MvCodeFragment) return Result.CONTINUE

        val phase = CompletionServiceImpl.getCompletionPhase()
        if (charTyped == ':') {
            if (phase is EmptyAutoPopup && phase.allowsSkippingNewAutoPopup(editor, charTyped)) {
                return Result.CONTINUE
            }

            AutoPopupController.getInstance(project).scheduleAutoPopup(editor)
            return Result.STOP
        }
        return Result.CONTINUE
    }
}
